core/schema/query.go

```diff
@@ -49,3 +57,53 @@ func (s *querySchema) pipeline(ctx *router.Context, parent *core.Query, args pip
 	})
 	return parent, nil
 }
+
+type checkVersionCompatibilityArgs struct {
+	Version string
+}
+
+func (s *querySchema) checkVersionCompatibility(ctx *router.Context, _ *core.Query, args checkVersionCompatibilityArgs) (bool, error) {
+	recorder := progrock.RecorderFromContext(ctx)
+
+	// Skip development version
+	if strings.Contains(engine.Version, "devel") {
+		recorder.Warn("Using development engine; skipping version compatibility check.")
+
+		return true, nil
+	}
+
+	engineVersion, err := semver.Parse(engine.Version)
+	if err != nil {
+		return false, err
+	}
+
+	sdkVersion, err := semver.Parse(args.Version)
+	if err != nil {
+		return false, err
+	}
+
+	// If the Engine is a major version above the SDK version, fails
+	// TODO: throw an error and abort the session
+	if engineVersion.Major > sdkVersion.Major {
+		recorder.Warn(fmt.Sprintf("Dagger engine version (%s) is significantly newer than the SDK's required version (%s). Please update your SDK.", engineVersion, sdkVersion))
+
+		// return false, fmt.Errorf("Dagger engine version (%s) is not compatible with the SDK (%s)", engineVersion, sdkVersion)
+		return false, nil
+	}
+
+	// If the Engine is older than the SDK, fails
+	// TODO: throw an error and abort the session
+	if engineVersion.LT(sdkVersion) {
+		recorder.Warn(fmt.Sprintf("Dagger engine version (%s) is older than the SDK's required version (%s). Please update your Dagger CLI.", engineVersion, sdkVersion))
+
+		// return false, fmt.Errorf("API version is older than the SDK, please update your Dagger CLI")
+		return false, nil
+	}
+
+	// If the Engine is a minor version newer, warn
+	if engineVersion.Minor > sdkVersion.Minor {
+		recorder.Warn(fmt.Sprintf("Dagger engine version (%s) is newer than the SDK's required version (%s). Consider updating your SDK.", engineVersion, sdkVersion))
+	}
```

Comment:
I think this should also return `false, nil`. You could have an SDK 3 minors lower and nothing's reported if using auto provisioning. Remember, even 1 v0.X minor can be incompatible.

Reply:
Okay, I'm doing it right now!

Reply:
What I'd actually do would be to start with:

```go
if engineVersion == sdkVersion.Minor {
    return true, nil
}
```
And assume `false` otherwise, just figuring the right warning message like you did. Since you're not returning errors and the Go and Node.js SDKs ignore the result, this boolean is only used so far in Python, in case of provisioning with `_EXPERIMENTAL_DAGGER_CLI_BIN`.